components/common_crawl_download/src/main.py

```diff
@@ -0,0 +1,161 @@
+            "warc_record_offset",
+            "warc_record_length",
+        ]
+
+        https_filesytem = HTTPFileSystem()
+
+        dataframe = dd.read_parquet(
+            self.index_files,
+            filters=self.filters,
+            columns=output_columns,
+            filesystem=https_filesytem,
+        )
+
+        return dataframe.set_index("url_surtkey", sorted=True, drop=True)
+
+    async def download_warc_content(self, row: t.Any) \
+            -> t.Tuple[str, str, t.Optional[str]]:
+        """
+        Download content of a single web page.
+
+        Args:
+            row: This should be a NamedTuple returned by df.itertuples(), but cannot be
+                 typehinted as such.
+
+        Returns:
+            A tuple containing the index, url, and extracted content
+        """
+        url = f"{CC_BASE_URL}/{row.warc_filename}"
+        headers = {
+            "Range": f"bytes={row.warc_record_offset}-"
+                     f"{row.warc_record_offset + row.warc_record_length - 1}",
+        }
+
+        try:
+            async with httpx.AsyncClient() as client:
+                response = await client.get(url, headers=headers)
+        except Exception as e:
+            logger.warning(f"Error downloading {url} with headers {headers}: {e}")
+            return row.Index, url, None
+        else:
+            warc_stream = io.BytesIO(response.content)
+            content = read_warc_content(warc_stream)
+
+            if self.extract_plain_text and content is not None:
+                content = extract_html(content)
+
+            return row.Index, row.url, content
+
+    def download_and_extract(self, dataframe: pd.DataFrame) -> pd.DataFrame:
+        """Concurrently download and extract the WARC files referenced in the provided dataframe."""
+        html_content = []
+
+        async def download_dataframe() -> None:
+            html = await asyncio.gather(
+                *[self.download_warc_content(row)
+                  for row in dataframe.itertuples()],
+            )
+            html_content.extend(html)
+
+        asyncio.run(download_dataframe())
+
+        columns = ["url_surtkey", "url", "content"]
+        if html_content:
+            content_df = pd.DataFrame(html_content, columns=columns)
+        else:
+            content_df = pd.DataFrame(columns=columns)
+
+        content_df = content_df.dropna()
+        content_df = content_df.set_index("url_surtkey", drop=True)
+
+        return content_df
+
+    def load(self) -> dd.DataFrame:
+        index_ddf = self.load_index()
+
+        if self.n_records_to_download is not None:
+            n_partitions = max(os.cpu_count(), self.n_records_to_download // 10000)  # type: ignore
+            logging.info(f"Repartitioning to {n_partitions} partitions.")
+            index_ddf = dd.from_pandas(index_ddf.head(self.n_records_to_download,
+                                                      npartitions=-1), npartitions=n_partitions)
```

Comment:
`npartitions=-1` Would that limit n_records_to_download to the size of the last partition?

Reply:
[It uses all partitions](https://docs.dask.org/en/latest/generated/dask.dataframe.DataFrame.head.html).

Have to admit that I don't know what the impact on performance is.